Updated and tested Seldon Core 1.15.0

axsaucedo opened the pull request:
Signed-off-by: Alejandro Saucedo <axsauze@gmail.com>

**Which issue is resolved by this Pull Request:**

Resolves part of #2311 addressing #2322.

**Description of your changes:**

* Adding requirements as per [Guidelines for /contrib Components](https://github.com/kubeflow/manifests/blob/master/proposals/20220926-contrib-component-guidelines.md)
* Fully tested kubeflow installation of Seldon Core's latest version 1.15.0
* Adding UPGRADE.md
* Adding automated tests (via `make -C contrib/seldon/ test`)
* Adding example testing for manual interaction

## Testing

A section was added under `contrib/seldon/README.md` containing the instructions to test, which are below.

It is now possible to run the automated tests with `make -C contrib/seldon/ test`.

## Explicit overview of manual install & testing 

Installation can be done with the kustomize command:

```
kustomize build seldon-core-operator/base | kubectl apply -n kubeflow -f -
```


We can create a test model once the "Install Seldon Operator" is configured

```
# Create namespace for model
kubectl create namespace seldon
```

We can create an echo model with the following command:

```
kubectl apply -f - << ENDapiVersion: machinelearning.seldon.io/v1
kind: SeldonDeployment
metadata:
  name: echo
  namespace: seldon
spec:
  predictors:
  - name: default
    replicas: 1
    graph:
      name: classifier
      type: MODEL
    componentSpecs:
    - spec:
        containers:
        - image: seldonio/echo-model:1.15.0-dev
          name: classifier
END
```

We can verify that model is running:

```
kubectl get pods -n seldon

NAME                                         READY   STATUS    RESTARTS   AGE
echo-default-0-classifier-679cb5fb68-qd4nm   2/2     Running   0          25m
```

Also we can verify that the correct virtualservice was created:

```
kubectl get virtualservice -n seldon

NAME   GATEWAYS                        HOSTS   AGE
echo   ["kubeflow/kubeflow-gateway"]   ["*"]   42m
```

Finally we can send a request (you will need to fetch the Dex Auth Token / Cookie):

```
export CLUSTER_IP=# Your cluster IP
export SESSION=# Your dex session

curl -H "Content-Type: application/json" -H "Cookie: authservice_session=${SESSION}" \
   -d '{"data": {"ndarray":[[1.0, 2.0, 5.0]]}}' \
   http://{CLUSTER_IP}/seldon/seldon/echo/api/v1.0/predictions

{"data":{"names":["t:0","t:1","t:2"],"ndarray":[[1.0,2.0,5.0]]},"meta":{"metrics":[{"key":"mycounter","type":"COUNTER","value":1},{"key":"mygauge","type":"GAUGE","value":100},{"key":"mytimer","type":"TIMER","value":20.2}],"requestPath":{"classifier":"seldonio/echo-model:1.15.0-dev"}}}
```

axsaucedo commented:
Thank you @juliusvonkohout - I have signed the CLI a while back (the https://cla.developers.google.com/clas confirms I have valid CLA), it seems the job is still "waiting for status", and I believe I saw it pass in a previous commit, is there a way to re-trigger the check?

juliusvonkohout commented:
/retest

annajung reviewed (commented):
Thanks for the prompt response in getting the component updated to work with the latest kubeflow release!

annajung commented on contrib/seldon/Makefile:
```diff
@@ -10,17 +10,35 @@
 clean-kustomize:
 	rm -rf seldon-core-operator/base
+
+test:
```
@kimwnasptd I think this should be sufficient to meet the test requirement for contrib, do we want to encourage to add the test into github actions workflow as well?

kimwnasptd replied:
I'd propose that yes, let's have a very minimal test in place for the GH Actions. This will allow us to run these tests periodically to easily smoke test these components

kimwnasptd replied:
@axsaucedo could you add a simple test workflow `.github/workflows/contrib_seldon_core.yaml` and define a minimal test case?

It can be something as simple as just verifying that the Pods of seldon core become ready. You can use as reference the rest of the apps that we are testing https://github.com/kubeflow/manifests/blob/master/.github/workflows/jwa_kind_test.yaml

If you want to test slightly more advanced cases, like creating a CR and testing if it also becomes Ready then that would also be more that welcome! But as a firs step testing that the core Pods are up should be enough

kimwnasptd replied:
Actually, just running this `make test` inside of the GH Action should suffice

axsaucedo replied:
Ok sounds good, let me add that

juliusvonkohout commented:
I think Seldon also has proper support for PodSecurityStandards restricted. At least that was the case for the previous version. Nevertheless that is something that should also be tested.

axsaucedo commented on contrib/seldon/UPGRADE.md:
```diff
@@ -0,0 +1,5 @@
+# Upgrading Documentation
```
Ok good point, I have added this in the upgrading docs including an env var so it can be done intuitively

axsaucedo commented on contrib/seldon/README.md:
```diff
@@ -14,5 +14,75 @@ kustomize build seldon-core-operator/base | kubectl apply -n kubeflow -f -
 This kustomize spec was created from the seldon-core-operator helm chart with:
```
Ok I've now added a one-liner explaining what it does, let me know your thoughts

axsaucedo commented on contrib/seldon/README.md:
````diff
@@ -14,5 +14,75 @@ kustomize build seldon-core-operator/base | kubectl apply -n kubeflow -f -
+## Testing
+
+```
+make test
````
My apologies, it seems the latest changes weren't updated, it should now work correctly

annajung replied:
verify that it works - thanks!

axsaucedo commented on contrib/seldon/Makefile:
```diff
@@ -10,17 +10,35 @@
-	sed -i "s#'seldon-webhook-service.kubeflow.svc.cluster.local'#"'$$(SERVICE_NAME).$$(SERVICE_NAMESPACE).svc.cluster.local#' seldon-core-operator/base/resources.yaml
-	sed -i "s#commonName: 'seldon-webhook-service.kubeflow.svc'#"'commonName: $$(SERVICE_NAME).$$(CERTIFICATE_NAMESPACE).svc#' seldon-core-operator/base/resources.yaml
-	sed -i "s#'seldon-webhook-service.kubeflow.svc'#"'$$(SERVICE_NAME).$$(SERVICE_NAMESPACE).svc#' seldon-core-operator/base/resources.yaml
+	sed -i "s#'seldon-webhook-service.kubeflow.svc.cluster.local'#"'seldon-webhook-service.$$(SERVICE_NAMESPACE).svc.cluster.local#' seldon-core-operator/base/resources.yaml
```
Ok thanks for the heads up, I've now added explicit dependency requirements in the UPGRADING.md, and referenced it in the README.md.

axsaucedo commented:
Ok it seems the CLA is passing now @juliusvonkohout.

![image](https://user-images.githubusercontent.com/1447507/203645497-eea70205-c071-4acb-8e08-29a28b4841b9.png)


@annajung thank you for the comments, these should now be resolved - let me know if anything else required.

axsaucedo commented:
@kimwnasptd ok I have now added the test - for some strange reason I can't seem to get the "Create KinD Cluster" step to pass in the CI, it seems it doesn't find the file `tests/gh-actions/kind-cluster-1-22.yaml` - have you seen this before?

kimwnasptd commented:
@axsaucedo we recently changed slightly our files for installing KinD https://github.com/kubeflow/manifests/pull/2331.

I'd suggest to do a rebase to the latest code and use the [`tests/gh-actions/kind-cluster.yaml`](https://github.com/kubeflow/manifests/blob/master/tests/gh-actions/kind-cluster.yaml) file for setting up the KinD cluster.

axsaucedo commented:
Thanks @kimwnasptd I've now rebased and updated, but now I'm getting `./tests/gh-actions/kind-cluster.yaml: Permission denied` - is there any permission updates that need to be set up in order for this to run, or is it an incorrect path?

kimwnasptd commented on .github/workflows/seldon_kind_test.yaml:
```diff
@@ -0,0 +1,29 @@
+      run: ./tests/gh-actions/install_kind.sh
+
+    - name: Create KinD Cluster
+      run: ./tests/gh-actions/kind-cluster.yaml
```
@axsaucedo I believe our issue is in this line, we'll need to run kind and use that file for creating the cluster.

The command should be
```bash
kind create cluster --config tests/gh-actions/kind-cluster.yaml
```
Similarly to https://github.com/kubeflow/manifests/blob/master/.github/workflows/jwa_kind_test.yaml#L20

axsaucedo replied:
Oh my mistake, not sure how but I assumed this was an `.sh` script, but definitely need to grab a coffee

kimwnasptd replied:
Can definitely relate to that

kimwnasptd commented on contrib/seldon/README.md:
```diff
@@ -11,8 +13,74 @@ kustomize build seldon-core-operator/base | kubectl apply -n kubeflow -f -
 ## Updating
 
-This kustomize spec was created from the seldon-core-operator helm chart with:
+See [UPGRADING.md](UPGRADING.md)
```
We'll need to rename this here to `UPGRADE.md`, since now it will redirect to a 404 page

axsaucedo replied:
Good catch! I've updated now

axsaucedo commented:
@kimwnasptd it seems I've managed to get everything working locally, however I still get an error with a missing CRD, it seems that HPA CRD for v2beta1 is not set up - is this expected? Should I be looking to set it up as part of the Seldon tests?

The error in the operator is:

```
no matches for kind \"HorizontalPodAutoscaler\" in version \"autoscaling/v2beta1\
```

kimwnasptd commented:
I think the problem is in this line in the logs https://github.com/kubeflow/manifests/actions/runs/3549119873/jobs/5961121555
```
{"level":"error","ts":1669390311.322579,"logger":"controller.seldon-controller-manager","msg":"Reconciler error","reconciler group":"machinelearning.seldon.io","reconciler kind":"SeldonDeployment","name":"echo","namespace":"seldon","error":"no matches for kind \"HorizontalPodAutoscaler\" in version \"autoscaling/v2beta1\"","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
```

I see that in K8s 1.22 The `HorizontalPodAutoscaler` had also the version `v2beta1`, but in K8s 1.25 they only have `v2beta2`
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#horizontalpodautoscaler-v1-autoscaling
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#horizontalpodautoscaler-v1-autoscaling

axsaucedo commented:
@kimwnasptd it seems that we may need to update some of the KIND tests as currently Seldon supports Kubernetes clusters up to 1.24 - currently we are in the process of upgrading the components, particularly HPA as per https://github.com/SeldonIO/seldon-core/pull/4172, however  have been careful as it would introduce breaking changes. Is there a way to set up the KIND cluster with kubernetes 1.24? I could add an extra config file to provide a pre-1-25-k8s kind test. Thank you  - let me know if you need more details.

axsaucedo commented:
@kimwnasptd thanks again for having a look, let me know your thoughts on the suggestion above, it does seem like the only way from our current state is to run the integration tests on a pre-1.25 k8s cluster, I can have a look at adding a `pre-1-24-...` kind setup config file if you think this can work (and add a note that emphasises the work + timeframe for post-1.25 support.

kimwnasptd commented:
@axsaucedo we discussed it with the release team as well in today's meeting (link to [notes](https://docs.google.com/document/d/1Cs9XYl_SHfGP9vs1eHyqb6JAcpE3ilM-5iRuBnV9AB4/edit#heading=h.qbxr0jfrv0f)).

In the KF 1.7 release we are targeting for K8s 1.25, but since we expect some vendors to not yet support this version we are also supporting K8s 1.24.

I'm OK with having this component working with K8s 1.24, but let's have a clear text somewhere at the top of the README that exposes the K8s version. So that it will be clear to users that will use KF 1.7 that this component will not work with 1.25 (to avoid confusion, since we will be promoting that the target K8s version for KF 1.7 will be K8s 1.25)

cc @annajung @DomFleischmann @jbottum

kimwnasptd commented:
So @axsaucedo to merge this PR could you:
1. Create a `tests/gh-actions/kind-cluster-1-24.yaml` configuration file for KinD and K8s 1.24
2. Update the GH Action for Seldon to use that yaml instead
3. Update the README to note that this works for K8s until 1.24

axsaucedo commented:
Ok perfect @kimwnasptd - very much appreciated. I will look at actioning these three points. Also absolutely agree on making a clear point in the README, as we also have a clear action plan towards ensuring seldon as a whole supports k8s 1.25+, we already have that PR suggested so it's a matter on how we can make sure we provide users with migration paths before we add it. 

In that case we'll add these changes as suggested and re-trigger the tests - I'll let you know if there are any issues / further questions. Thanks.

annajung commented on .github/workflows/seldon_kind_test.yaml:
```diff
@@ -0,0 +1,32 @@
+name: Build & Apply Seldon manifests in KinD
+on:
+  pull_request:
+    paths:
+      - contrib/seldon/**
```
I think this should be okay for now, but @kimwnasptd we should think about other ways to trigger `/contrib` tests in the future and not just trigger based on changes to the `contrib/seldon` directory.

kimwnasptd replied:
Was thinking the same thing! We could probably use the `workflow_dispatch` event as well in these workflows, to be able to manually trigger them https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch

Let's play around with this more once we have the components update their manifests and have some basic tests in place.

annajung commented on contrib/seldon/README.md:
```diff
@@ -1,6 +1,14 @@
+
+* Kubernetes 1.18 - 1.24
+
+Support for Kubernetes 1.25 is currently part of (SeldonIO/seldon-core#4172)[https://github.com/SeldonIO/seldon-core/pull/4172]
```
```suggestion
Support for Kubernetes 1.25 is currently part of [SeldonIO/seldon-core#4172](https://github.com/SeldonIO/seldon-core/pull/4172)
```

axsaucedo replied:
Good catch - thank you for the commit suggest

axsaucedo commented on README.md:
````diff
@@ -274,7 +274,15 @@ Install the Katib official Kubeflow component:
 ```
 
-#### Central Dashboard
+#### Seldon
````
Ok that makes sense, I have now reverted it to what it was

annajung commented:
Thanks for working with us to update Seldon to meet the contrib guidelines
/lgtm

axsaucedo commented:
Fantastic @annajung - thanks for the update, let me know if anything else needed fromy my side to merge 👍

kimwnasptd commented:
Was about to comment regarding the README.md, but @annajung had covered it already https://github.com/kubeflow/manifests/pull/2326#discussion_r1035086822, nice!

@axsaucedo thank you for your time and great work on updating these! The tests will also help us ensuring the quality of the manifests we bring along and even raise the standard.

/lgtm
/approve

umka1332 commented on example/kustomization.yaml:
```diff
@@ -38,7 +38,7 @@ resources:
 # Notebook Controller
 - ../apps/jupyter/notebook-controller/upstream/overlays/kubeflow
 # Profiles + KFAM
-- ../apps/profiles/upstream/overlays/kubeflow
+# - ../apps/profiles/upstream/overlays/kubeflow
```
Are there any reason to remove Profile controller from Kubeflow example installation?

juliusvonkohout commented:
@axsaucedo  why did you delete kubeflow-edit-seldon.yaml?
